File: docs/video/tools/assemble.py
#!/usr/bin/env python3
"""
Assemble the POLARIS-X demo video:
  title card → recorded take (warm-up trimmed) → end card,
  with the narration laid on at the recorded cue times, plus an .srt.

Run:  python3 /tmp/vidgen/assemble.py
"""
import json, os, re, subprocess, sys

FF = open('/tmp/vidgen/ff_path.txt').read().strip()
W = '/tmp/vidgen'
OUT = '/home/user/dakshin-marg/docs/video'
AUDIO = f'{OUT}/audio'
os.makedirs(f'{W}/build', exist_ok=True)
os.makedirs(OUT, exist_ok=True)

TITLE_DUR = 4.5
END_DUR = 7.0
FPS = 30


def run(args, label):
    p = subprocess.run(args, capture_output=True)
    if p.returncode != 0:
        print(f'FAIL {label}\n' + p.stderr.decode()[-2500:])
        sys.exit(1)
    print(f'ok   {label}')


tl = json.load(open(f'{W}/timeline.json'))
cues = {c['name']: c['t'] for c in tl['cues']}
take = cues['take_start']
total = tl['total']
take_dur = total - take + 3.0           # let the last frame settle
print(f'take starts at {take:.2f}s · take length {take_dur:.2f}s')

# narration placement: clip → (cue name, offset seconds inside the act)
PLAN = [
    ('s01_hook',            'shot01_cue',        0.4),
    ('s02_live',            'shot02_cue',        0.4),
    ('s03_seaice',          'shot03_cue',        0.4),
    ('s04_icebergs',        'shot04_cue',        0.6),
    ('s05_risk',            'shot05_cue',        0.4),
    ('s06_routes',          'shot06_cue',        0.4),
    ('s07_mission_create',  'shot07_cue',        0.6),
    ('s08_underway',        'shot08_cue',        0.4),
    ('s09_review_replan',   'shot09_cue',        0.4),
    ('s10_drill',           'shot10_cue',        0.4),
]

# ── narration audio: one delayed copy per clip, mixed ────────────────────────
inputs, filters, mixes = [], [], []
for i, (clip, cue, off) in enumerate(PLAN):
    path = f'{AUDIO}/{clip}.mp3'
    if not os.path.exists(path):
        print('missing', path); sys.exit(1)
    inputs += ['-i', path]
    delay = int(round((TITLE_DUR + (cues[cue] - take) + off) * 1000))
    filters.append(f'[{i}:a]adelay={delay}|{delay},volume=1.0[a{i}]')
    mixes.append(f'[a{i}]')
filters.append(''.join(mixes) + f'amix=inputs={len(PLAN)}:normalize=0:duration=longest[out]')
run([FF, '-y', '-hide_banner', '-loglevel', 'error', *inputs,
     '-filter_complex', ';'.join(filters), '-map', '[out]',
     '-c:a', 'aac', '-b:a', '192k', '-ar', '48000',
     f'{W}/build/narration.m4a'], 'narration mix')

# ── video segments ──────────────────────────────────────────────────────────
run([FF, '-y', '-hide_banner', '-loglevel', 'error', '-loop', '1', '-t', str(TITLE_DUR), '-i', f'{W}/cards/title.png',
     '-vf', f'scale={1600}:{900},fps={FPS},fade=t=in:st=0:d=0.7,fade=t=out:st={TITLE_DUR-0.7}:d=0.7,format=yuv420p',
     '-c:v', 'libx264', '-preset', 'medium', '-crf', '19', f'{W}/build/01_title.mp4'], 'title card')

import glob
cands = sorted(glob.glob(f'{W}/video/*.webm'), key=os.path.getmtime)
if not cands:
    print('no recorded webm in', f'{W}/video'); sys.exit(1)
TAKE_FILE = cands[-1]
print('take file:', TAKE_FILE)
run([FF, '-y', '-hide_banner', '-loglevel', 'error', '-ss', f'{max(0, take - 1.0):.2f}', '-i', TAKE_FILE,
     '-t', str(take_dur), '-vf', f'fps={FPS},scale=1600:900,format=yuv420p',
     '-c:v', 'libx264', '-preset', 'medium', '-crf', '19', '-an', f'{W}/build/02_take.mp4'], 'take segment')

run([FF, '-y', '-hide_banner', '-loglevel', 'error', '-loop', '1', '-t', str(END_DUR), '-i', f'{W}/cards/end.png',
     '-vf', f'scale={1600}:{900},fps={FPS},fade=t=in:st=0:d=0.7,fade=t=out:st={END_DUR-0.9}:d=0.9,format=yuv420p',
     '-c:v', 'libx264', '-preset', 'medium', '-crf', '19', f'{W}/build/03_end.mp4'], 'end card')

with open(f'{W}/build/list.txt', 'w') as fh:
    for seg in ('01_title', '02_take', '03_end'):
        fh.write(f"file '{W}/build/{seg}.mp4'\n")
run([FF, '-y', '-hide_banner', '-loglevel', 'error', '-f', 'concat', '-safe', '0', '-i', f'{W}/build/list.txt',
     '-c', 'copy', f'{W}/build/silent.mp4'], 'concat')

# ── mux ─────────────────────────────────────────────────────────────────────
probe = subprocess.run([FF, '-i', f'{W}/build/silent.mp4'], capture_output=True).stderr.decode()
dur = float(re.search(r'Duration: (\d+):(\d+):([\d.]+)', probe).group(1)) * 3600 \
    + float(re.search(r'Duration: (\d+):(\d+):([\d.]+)', probe).group(2)) * 60 \
    + float(re.search(r'Duration: (\d+):(\d+):([\d.]+)', probe).group(3))
print(f'video duration {dur:.1f}s')
run([FF, '-y', '-hide_banner', '-loglevel', 'error', '-i', f'{W}/build/silent.mp4', '-i', f'{W}/build/narration.m4a',
     '-filter_complex', f'[1:a]apad,atrim=0:{dur:.2f}[a]', '-map', '0:v', '-map', '[a]',
     '-c:v', 'copy', '-c:a', 'aac', '-b:a', '192k', '-movflags', '+faststart',
     f'{OUT}/POLARIS-X-demo.mp4'], 'final mux')
run([FF, '-y', '-hide_banner', '-loglevel', 'error', '-i', f'{W}/build/narration.m4a',
     '-t', f'{dur:.2f}', '-c:a', 'libmp3lame', '-b:a', '192k', f'{OUT}/POLARIS-X-narration.mp3'], 'narration mp3')

# ── subtitles ───────────────────────────────────────────────────────────────
TEXTS = {
    's01_hook': "Every austral summer, India's research stations are resupplied through Antarctic sea ice — where the ice charts are days old, icebergs move, and the ice edge can shift overnight. POLARIS-X turns open satellite and model data into a risk picture, route options, and a re-planning decision — and it keeps the human in command at every step.",
    's02_live': "This is the running system, in LIVE mode. Everything on screen is computed from real datasets — NSIDC sea-ice concentration, USNIC iceberg positions, BYU and NIC iceberg tracks, and Open-Meteo wind and temperature. Every layer declares what it is and how old it is: observation, model output, or planned. Ocean currents and waves are planned — no data source is connected yet, so the system says so, instead of drawing something that does not exist.",
    's03_seaice': "This is the latest NSIDC daily analysis in our data window — the first of September, mean concentration thirty-nine point eight percent along the corridor. I can step back through eight observation days, or forecast forward. At plus forty-eight hours the damped-trend model gives forty-two percent, with an uncertainty band of five point seven percent and a backtest error of four percent — against four point two for persistence. The badge changes to SEA-ICE FORECAST, because this is model output, not an observation.",
    's04_icebergs': "Thirteen icebergs are charted in this area; seventeen have real tracks from the BYU database, and five are classified as moving. The drift forecast carries an empirical P90 corridor of about forty-two kilometres at seven days. And the card states openly that the track archive lags real time — which is why current positions come from the USNIC chart.",
    's05_risk': "Risk is the core product: the IMO POLARIS risk index for sea ice, Overland's icing predictor, and iceberg hazard zones — combined worst-case, never averaged. With a PC5 ice class, the worst cell in the area is normal operation: RIO plus ten. Now watch the only thing I change — the vessel. PC7: minus nine, and a quarter of the cells move into the elevated band. No ice class at all: minus forty-nine, and over half the area becomes critical. Same ocean, same data — the vessel changes the answer.",
    's06_routes': "For routing, the optimizer does not invent cost weights — it applies severity ceilings as hard constraints. DIRECT is eight hundred and sixty-seven nautical miles, sixty-nine hours, with twenty percent of the route in medium severity. CONSERVATIVE costs a hundred and thirty extra miles to keep the whole route low. The recommendation cites IMO circular 1519 — avoid elevated-risk areas when planning. And fuel reads NOT COMPUTED, because no validated consumption model exists. We do not guess numbers we cannot defend.",
    's07_mission_create': "Now the real workflow — a resupply mission. Origin picked on the map, destination Bharati station in Prydz Bay, departure the second of September. The wizard validates the date against the actual data window — the twenty-fifth of August to the fourth of September — and tells me which observation and which weather hour it will use. Ice class PC5, twelve and a half knots, and a deliberately strict ceiling: nothing worse than low on the remaining route.",
    's08_underway': "Three options for that departure — all eight hundred and sixty-nine miles and low at this hour. I take the recommended DIRECT line. Acceptance is a human click: the system never starts a voyage by itself. The vessel then moves along the accepted geometry at the optimizer's own speed model — there is no decorative animation. I advance the mission clock twelve hours at a time; conditions at the vessel update with the clock, and every three simulated hours the remaining route is re-assessed against the data valid at that time.",
    's09_review_replan': "There it is — ROUTE REVIEW REQUIRED. The re-assessment found medium severity on the remaining route, which breaks the ceiling the operator set. It names the primary factor, and it pauses. It does not reroute on its own. I ask for alternatives from our current position, at the current time. Three candidates; CONSERVATIVE is recommended — five hundred and seventy miles, and the smallest exposure of the three. I accept it: the active route changes, the track restarts from this point, and the decision is written into the mission log.",
    's10_drill': "For the finale, a repeatable drill. It simulates exactly one fact — iceberg D23 ungrounding and drifting a hundred kilometres north, into our corridor — and every route, every risk number and every corridor radius is computed by the same live engines you have just seen. Twenty-four hours at cruise speed: three hundred miles covered, then the simulated re-sighting arrives, and D23 is inside the corridor. The remaining leg goes from low to critical, with a closest approach of two and a half kilometres. Re-planning recommends CONSERVATIVE: five hundred and ninety-two miles, all low — twenty-five miles longer than staying put. I accept it, and the new route becomes active.",
}

def ts(sec):
    h, rem = divmod(sec, 3600); m, s = divmod(rem, 60)
    return f'{int(h):02d}:{int(m):02d}:{s:06.3f}'.replace('.', ',')

def chunks(text, n):
    words = text.split()
    size = max(1, len(words) // n + (1 if len(words) % n else 0))
    return [' '.join(words[i:i + size]) for i in range(0, len(words), size)]

lines, idx = [], 1
for clip, cue, off in PLAN:
    start = TITLE_DUR + (cues[cue] - take) + off
    # use the audio's real length for subtitle pacing
    p = subprocess.run([FF, '-i', f'{AUDIO}/{clip}.mp3'], capture_output=True).stderr.decode()
    mm = re.search(r'Duration: (\d+):(\d+):([\d.]+)', p)
    alen = float(mm.group(1)) * 3600 + float(mm.group(2)) * 60 + float(mm.group(3))
    txt = TEXTS[clip]
    parts = chunks(txt, max(2, round(alen / 7)))
    total_chars = sum(len(x) for x in parts)
    t = start
    for part in parts:
        d = alen * (len(part) / total_chars)
        lines.append(f'{idx}\n{ts(t)} --> {ts(t + d - 0.05)}\n{part}\n')
        idx += 1
        t += d

open(f'{OUT}/POLARIS-X-demo.srt', 'w').write('\n'.join(lines))

print('\ncues (take-relative):')
for k in sorted([c for c in cues if c.startswith('shot') or c in ('take_start', 'end')]):
    print(f'  {k:24s} {cues[k] - take:7.2f}s')
print('final:', f'{OUT}/POLARIS-X-demo.mp4')

File: tools/demo-video/src/assemble.mjs
/**
 * Assembler — turn the recorded take + narration clips into the deliverables:
 *
 *   docs/video/Dakshin-Marg-demo.mp4         master (1280×720, 30 fps, narrated)
 *   docs/video/Dakshin-Marg-demo-720p.mp4    shareable cut (1280×720)
 *   docs/video/Dakshin-Marg-narration.mp3    voice track only
 *   docs/video/Dakshin-Marg-demo.srt         subtitles, paced from the real audio
 *
 * Structure: title card → recorded take (warm-up trimmed) → end card, with the
 * narration laid on at the cue times the recorder wrote into work/timeline.json.
 */
import { spawnSync } from 'node:child_process';
import { existsSync, mkdirSync, readdirSync, readFileSync, rmSync, statSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
// playwright-core is imported lazily so preflight can install it first
import { FILES, HERE, NARRATION_DIR, VIDEO, WORK, ensureDirs, makeLogger } from './config.mjs';
import { resolveFfmpeg } from './preflight.mjs';
import { probeDuration, spoken } from './tts.mjs';

const log = makeLogger();

function ff(args, label) {
  const bin = resolveFfmpeg({ install: false });
  if (!bin.ok) throw new Error(bin.reason);
  const r = spawnSync(bin.ffmpeg, ['-hide_banner', '-loglevel', 'error', '-y', ...args], { encoding: 'utf8', maxBuffer: 32 * 1024 * 1024 });
  if (r.status !== 0) {
    throw new Error(`ffmpeg failed (${label}):\n${(r.stderr || r.stdout || '').slice(-1800)}`);
  }
  return r;
}

function newestWebm(dir = join(WORK, 'video')) {
  if (!existsSync(dir)) return null;
  const files = readdirSync(dir).filter((f) => /\.(webm|mp4)$/i.test(f)).map((f) => join(dir, f));
  if (files.length === 0) return null;
  return files.sort((a, b) => statSync(b).mtimeMs - statSync(a).mtimeMs)[0];
}

/* ── title / end cards ──────────────────────────────────────────────────── */

export async function renderCards({ chromium: chr, width = VIDEO.width, height = VIDEO.height, script }) {
  const cardsDir = join(WORK, 'cards');
  mkdirSync(cardsDir, { recursive: true });
  let executablePath = chr.executablePath;
  if (chr.inflate) executablePath = await chr.inflate();
  const { chromium: chromiumPw } = await import('playwright-core');
  const browser = await chromiumPw.launch({ executablePath, headless: true, args: chr.args ?? [] });
  try {
    const page = await (await browser.newContext({ viewport: { width, height }, deviceScaleFactor: 1 })).newPage();
    for (const name of ['title', 'end']) {
      const html = join(HERE, 'cards', `${name}.html`);
      if (!existsSync(html)) throw new Error(`missing card template: ${html}`);
      let body = readFileSync(html, 'utf8')
        .replaceAll('{{PRODUCT}}', script.product ?? 'Dakshin Marg')
        .replaceAll('{{YEAR}}', String(new Date().getFullYear()));
      await page.setContent(body, { waitUntil: 'load' });
      await page.waitForTimeout(700);
      const out = join(cardsDir, `${name}.png`);
      await page.screenshot({ path: out });
      log.info(`  card ${name} → ${out}`);
    }
  } finally {
    await browser.close();
  }
  return { title: join(cardsDir, 'title.png'), end: join(cardsDir, 'end.png') };
}

/* ── subtitles ──────────────────────────────────────────────────────────── */

const srtTs = (sec) => {
  const s = Math.max(0, sec);
  const h = Math.floor(s / 3600);
  const m = Math.floor((s % 3600) / 60);
  const rest = s % 60;
  return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}:${rest.toFixed(3).padStart(6, '0')}`.replace('.', ',');
};

function chunkText(text, n) {
  const words = text.split(/\s+/).filter(Boolean);
  const size = Math.max(1, Math.ceil(words.length / Math.max(1, n)));
  const out = [];
  for (let i = 0; i < words.length; i += size) out.push(words.slice(i, i + size).join(' '));
  return out;
}

function buildSrt(placements) {
  const lines = [];
  let idx = 1;
  for (const p of placements) {
    const text = spoken(p.text);
    const parts = chunkText(text, Math.max(2, Math.round(p.duration / 6.5)));
    const chars = parts.reduce((a, b) => a + b.length, 0) || 1;
    let t = p.start;
    for (const part of parts) {
      const d = p.duration * (part.length / chars);
      lines.push(`${idx}\n${srtTs(t)} --> ${srtTs(t + d - 0.05)}\n${part}\n`);
      idx += 1;
      t += d;
    }
  }
  writeFileSync(FILES.subtitles, lines.join('\n'));
  return idx - 1;
}

/* ── the assembly ───────────────────────────────────────────────────────── */

export async function assemble({ script, chromium: chr, share = true, keepBuild = false } = {}) {
  ensureDirs();
  log.step('Assemble — cards, take, narration, subtitles');

  if (!existsSync(FILES.timeline)) throw new Error(`no recorded timeline at ${FILES.timeline} — run the record step first`);
  const tl = JSON.parse(readFileSync(FILES.timeline, 'utf8'));
  const cues = Object.fromEntries((tl.cues ?? []).map((c) => [c.name, c.t]));
  const takeStart = cues.take_start;
  if (takeStart === undefined) throw new Error('timeline has no take_start cue');
  const takePad = Number(process.env.DM_TAKE_PAD ?? 0);   // extra seconds trimmed off the take head

  const take = (tl.takeFile && existsSync(tl.takeFile)) ? tl.takeFile : newestWebm();
  if (!take) throw new Error(`no recording in ${join(WORK, 'video')} — run the record step first`);
  log.info(`take: ${take} (${(statSync(take).size / 1e6).toFixed(1)} MB)`);

  const durations = tl.durations ?? {};
  const missing = script.acts.filter((a) => !existsSync(join(NARRATION_DIR, `${a.id}.mp3`)));
  if (missing.length) log.warn(`no audio for: ${missing.map((a) => a.id).join(', ')} (these acts stay silent)`);

  /* cards */
  const cards = await renderCards({ chromium: chr, script, width: 1600, height: 900 });  // cards are laid out for 1600×900; segments downscale
  const titleDur = Math.max(script.cards?.titleSeconds ?? 5, (durations['00_title'] ?? 0) + 1.6);

  /* narration placement */
  const OFFSET = 0.4;                                  // seconds after the cue
  const placements = [];
  for (const act of script.acts) {
    const file = join(NARRATION_DIR, `${act.id}.mp3`);
    if (!existsSync(file)) continue;
    const dur = durations[act.id] ?? probeDuration(file);
    let start;
    if (act.id === '00_title') start = 0.35;
    else {
      const cue = cues[`${act.id}_cue`];
      if (cue === undefined) { log.warn(`act ${act.id} has no recorded cue — skipping its narration`); continue; }
      start = Math.max(titleDur + 0.3, titleDur + (cue - takeStart - takePad) + OFFSET);
    }
    placements.push({ id: act.id, file, start, duration: dur, text: act.text });
  }
  placements.sort((a, b) => a.start - b.start);

  /* narration mix */
  const inputs = [];
  const filters = [];
  placements.forEach((p, i) => {
    inputs.push('-i', p.file);
    const delay = Math.max(0, Math.round(p.start * 1000));
    filters.push(`[${i}:a]aresample=48000,adelay=${delay}|${delay},volume=1.0[a${i}]`);
  });
  const mixFile = join(WORK, 'build', 'narration.m4a');
  if (placements.length === 0) {
    log.warn('no narration clips — the video will be silent');
    ff(['-f', 'lavfi', '-i', 'anullsrc=r=48000:cl=stereo', '-t', '1', '-c:a', 'aac', mixFile], 'silent track');
  } else {
    filters.push(`${placements.map((_, i) => `[a${i}]`).join('')}amix=inputs=${placements.length}:normalize=0:duration=longest[out]`);
    ff([...inputs, '-filter_complex', filters.join(';'), '-map', '[out]',
      '-c:a', 'aac', '-b:a', '192k', '-ar', '48000', mixFile], 'narration mix');
    log.ok(`narration mix — ${placements.length} clip(s)`);
  }

  /* video segments */
  const takeDur = Math.max(1, (tl.total ?? 0) - takeStart + 2.5 - takePad);
  const buildDir = join(WORK, 'build');
  mkdirSync(buildDir, { recursive: true });

  ff(['-loop', '1', '-t', String(titleDur), '-i', cards.title,
    '-vf', `scale=${VIDEO.width}:${VIDEO.height},fps=${VIDEO.fps},fade=t=in:st=0:d=0.7,fade=t=out:st=${(titleDur - 0.7).toFixed(2)}:d=0.7,format=yuv420p`,
    '-c:v', 'libx264', '-preset', 'medium', '-crf', String(VIDEO.crf), '-an',
    join(buildDir, '01_title.mp4')], 'title card');

  const capFirst = tl.capture?.firstFrameMy ?? 0;   // take-file time of the session clock zero
  ff(['-ss', `${Math.max(0, takeStart - 1.0 - capFirst + takePad).toFixed(2)}`, '-i', take, '-t', `${takeDur.toFixed(2)}`,
    '-vf', `fps=${VIDEO.fps},scale=${VIDEO.width}:${VIDEO.height}:force_original_aspect_ratio=decrease,pad=${VIDEO.width}:${VIDEO.height}:(ow-iw)/2:(oh-ih)/2,format=yuv420p`,
    '-c:v', 'libx264', '-preset', 'veryfast', '-crf', String(VIDEO.crf), '-an',
    join(buildDir, '02_take.mp4')], 'take segment');

  // the end card must cover any narration that runs past the take
  const voiceEnd = placements.reduce((m, p) => Math.max(m, p.start + p.duration), 0);
  const videoBeforeEnd = titleDur + takeDur;
  const endDur = Math.max(script.cards?.endSeconds ?? 8, voiceEnd - videoBeforeEnd + 1.8, 4);
  ff(['-loop', '1', '-t', `${endDur.toFixed(2)}`, '-i', cards.end,
    '-vf', `scale=${VIDEO.width}:${VIDEO.height},fps=${VIDEO.fps},fade=t=in:st=0:d=0.7,fade=t=out:st=${(endDur - 0.9).toFixed(2)}:d=0.9,format=yuv420p`,
    '-c:v', 'libx264', '-preset', 'medium', '-crf', String(VIDEO.crf), '-an',
    join(buildDir, '03_end.mp4')], 'end card');

  const list = join(buildDir, 'list.txt');
  writeFileSync(list, ['01_title', '02_take', '03_end'].map((s) => `file '${join(buildDir, `${s}.mp4`)}'\n`).join(''));
  const silent = join(buildDir, 'silent.mp4');
  ff(['-f', 'concat', '-safe', '0', '-i', list, '-c', 'copy', silent], 'concat');

  const totalDur = titleDur + takeDur + endDur;
  ff(['-i', silent, '-i', mixFile,
    '-filter_complex', `[1:a]apad,atrim=0:${totalDur.toFixed(2)},asetpts=PTS-STARTPTS[a]`,
    '-map', '0:v', '-map', '[a]', '-c:v', 'copy', '-c:a', 'aac', '-b:a', '192k',
    '-shortest', '-movflags', '+faststart', FILES.master], 'final mux');
  log.ok(`master → ${FILES.master}`);

  if (share) {
    ff(['-i', FILES.master,
      '-vf', `scale=${VIDEO.shareWidth}:${VIDEO.shareHeight},format=yuv420p`,
      '-c:v', 'libx264', '-preset', 'veryfast', '-crf', String(VIDEO.shareCrf),
      '-c:a', 'aac', '-b:a', '160k', '-movflags', '+faststart', FILES.share], '720p share cut');
    log.ok(`shareable → ${FILES.share}`);
  }

  ff(['-i', mixFile, '-t', `${totalDur.toFixed(2)}`, '-c:a', 'libmp3lame', '-b:a', '160k', FILES.narrationTrack], 'narration track');

  const subs = buildSrt(placements);
  log.ok(`subtitles → ${FILES.subtitles} (${subs} cues)`);

  if (!keepBuild) {
    for (const f of ['01_title.mp4', '02_take.mp4', '03_end.mp4', 'silent.mp4']) {
      rmSync(join(buildDir, f), { force: true });
    }
  }

  const mb = (p) => (existsSync(p) ? `${(statSync(p).size / 1e6).toFixed(1)} MB` : '—');
  log.step('Done');
  log.info(`master      ${FILES.master}   ${mb(FILES.master)}`);
  log.info(`720p        ${FILES.share}   ${mb(FILES.share)}`);
  log.info(`narration   ${FILES.narrationTrack}   ${mb(FILES.narrationTrack)}`);
  log.info(`subtitles   ${FILES.subtitles}`);
  log.info(`runtime     ${fmtTime(totalDur)} (title ${titleDur.toFixed(1)}s + take ${takeDur.toFixed(1)}s + end ${endDur.toFixed(1)}s)`);
  return { master: FILES.master, share: FILES.share, subtitles: FILES.subtitles, totalDur, titleDur, takeDur, endDur };
}

export function fmtTime(sec) {
  const m = Math.floor(sec / 60);
  const s = Math.round(sec % 60);
  return `${m}:${String(s).padStart(2, '0')}`;
}
